developer-docs/user-proxy.md

```diff
@@ -0,0 +1,118 @@
+
+This problem is demonstrated in the below sequence diagram
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Service A
+    participant Service B
+    participant Keycloak
+
+    # user login
+    Note right of User: login
+    User->>Keycloak: login(creds)
+    Keycloak-->>User: user token
+
+    # user normal request
+    Note right of User: typical request
+    User->>Service A: POST /action [user token]
+    Service A->>Service A: actions
+    Service A-->>User: 200 OK
+
+    # problematic request
+    Note over User,Service A: problematic proxy request
+    User->>Service A: POST /action [user token]
+    Note right of Service A: service account login
+    Service A->>Keycloak: login(creds)
+    Keycloak-->>Service A: service token
+    Service A->>Service B: /special-route [service token]
+    Service B->>Service B: checks auth
+    Note over Service B: At this point, Service B can't access<br/>the original roles of the User.<br/> For example, what if the user is an admin.
+    Service B-->>Service A: 401 Not Authorised
+    Service A-->>User: 401 Not Authorised
+```
+
+Here are some ideas to remedy this
+
+1. add flags into the request payload which include overrides or other user auth information
+   - not ideal as this would mean ensuring all proxy payloads are changed to meet this need, and it also means user info is likely to leak in logs, not super maintainable
+1. add the user's token somewhere into the payload
+   - this would raise questions like whether we check the token expiry, also could leak user info (since the token is not encrypted), it also means a user could more easily spoof a valid inter-service request if they somehow get the correct role
+1. lookup the user's information dynamically based on minimum information such as the user's username
+   - this is not a bad solution, but it is not very performant (as we need to do a network request to the keycloak server), and it would also require granting admin type access for the keycloak API to all the services which might need to know about the user - this is violating our plan for JWT's being sufficient information to validate a user request without additional external requests
+
+Given the above, an approach where we a) don't need to interact with the Keycloak server b) don't need to uniquely add payload information for all the relevant info on all the models and c) don't leak user details to logs where possible is ideal. The solution implemented is below.
+
+## Implemented approach
+
+---
+
+**Key solution**: when service A makes a service-only request to service B, it uses a symmetric encryption key to encrypt the user's info - this is placed on a header `X-User-Context`. Service B decrypts this user context and can use it, where needed, to make decisions about whether an action should be allowed, or other determinations.
```

Comment:
Cool - as i was reading the motivating problem statements an idea like this to somehow pass on user context came to mind. I think the symmetric encryption key is a great idea to help facilitate this.

Reply:
Great